File: Operators/cross_sim_binary.m
function [output1,output2] = cross_sim_binary(varargin)
% Simulated binary crossover.

mode = varargin{end};
switch mode
    case 'execute'
        Parent = varargin{1};
        Para   = varargin{3};

        Parent = Parent.decs;
        DisC = Para;
            
        [N,D] = size(Parent);
        Parent1 = Parent(1:ceil(N/2),:);
        Parent2 = Parent(floor(N/2)+1:end,:);
        Nhalf = size(Parent1,1);

        beta = zeros(Nhalf,D);
        mu = rand(Nhalf,D);

        beta(mu<=0.5) = (mu(mu<=0.5)*2).^(1/(1+DisC));
        beta(mu>0.5)  = (2-mu(mu>0.5)*2).^(-1/(1+DisC));
        Offspring = [0.5*((1+beta).*Parent1+(1-beta).*Parent2);
            0.5*((1-beta).*Parent1+(1+beta).*Parent2)];        
        output1   = Offspring(1:N,:);
        output2   = varargin{5};

    case 'parameter'
        output1 = [20,40]; % crossover distribution

    case 'behavior'
        output1 = {'';'GS'}; % always perform global search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/cross_sim_binary.py
"""Simulated binary crossover (SBX) operator."""

from __future__ import annotations

from typing import Any

import numpy as np


def cross_sim_binary(*args: Any):
    mode = args[-1]
    if mode == "execute":
        parent = args[0]
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None

        decs = getattr(parent, "decs", None)
        if callable(decs):
            parent = decs()
        else:
            parent = decs if decs is not None else parent
        parent = np.asarray(parent, dtype=float)
        n, d = parent.shape
        first = parent[: (n + 1) // 2]
        second = parent[n // 2 :]
        n_half = first.shape[0]

        eta = float(np.asarray(para, dtype=float).reshape(-1)[0]) if para is not None else 20.0
        mu = np.random.rand(n_half, d)
        beta = np.empty_like(mu)
        mask = mu <= 0.5
        beta[mask] = (2.0 * mu[mask]) ** (1.0 / (1.0 + eta))
        beta[~mask] = (2.0 - 2.0 * mu[~mask]) ** (-1.0 / (1.0 + eta))

        offspring = np.vstack(
            (
                0.5 * ((1.0 + beta) * first + (1.0 - beta) * second),
                0.5 * ((1.0 - beta) * first + (1.0 + beta) * second),
            )
        )
        return offspring[:n], aux

    if mode == "parameter":
        return np.array([20.0, 40.0], dtype=float), None

    if mode == "behavior":
        return ["", "GS"], None

    raise ValueError(f"Unsupported mode: {mode}")

